the-collab-lab/tcl-27-smart-shopping-list
# Tc dp join shopping list

Co-authored-by: Dani Poplin <danih37@gmail.com>

## Description

This PR updates our data structure, adds a new Home.js component, updates GroceryCard.js, and updates AddItem.js.

#### Data Structure

We updated Firestore database such that the item lists are created under a user token document, in the items collection. This allows us to search the database for an individual user token and fetch the data object containing list items. 

#### Home.js

We created a new home component that allows the user to either create a new list or join an existing list by entering a user token. If the user wants to join an existing list, we confirm that the list already exists in the database, if so, we save the token to `localStorage`. We then query the database for the entered token, fetch the existing list items and redirect the user to the list view. 

#### GroceryCard.js

We updated the unordered list items to display the data based on the new data structure in the database. 

#### AddItem.js

We updated add item component to save the data to the database in the new structure and to clear the add item form upon submission. 

#### App.js

We updated the `Switch` statement such that the user is not redirected to the list view if the user is not logged in. 

## Related Issue

Closes #5 

## [Acceptance Criteria](https://github.com/the-collab-lab/tcl-27-smart-shopping-list/issues/5)

- When the user does not already have a token in localStorage, on the onboarding/home screen, a simple form is displayed that allows the user to enter a token
- Entering the token and hitting submit saves the token to localStorage, effectively giving them joint control of the list
- On submit, show an error if the token does not exist
- If they get an error message, allow them to try again or create a new list

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|  ✓ | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |


## Testing Steps / QA Criteria

1. First, from your terminal, pull from the `main` branch with `git pull origin main`
2. From your terminal, pull down this branch with `git pull origin tc-dp-join-shopping-list` and check that branch out with `git checkout tc-dp-join-shopping-list`
3. Ensure that a simple form is displayed on the home screen to allow the user to enter a token if the user is not already logged in. For this step, run `localStorage.clear()` in your console. 
4. Go to the [Firestore database](https://console.firebase.google.com/u/0/project/tcl-27-shopping-list/firestore/data/~2Fgroceries) and copy one of the tokens listed in the collection.
5. Enter the copied token in the form on the home page and hit submit. 
6. Check the the same token is saved to your `localStorage` by running `localStorage` in your console. 
7. On submit, you should be redirected to the list view.
8. Clear your `localStorage` by running `localStorage.clear()` in your console.
9. Enter a random word in the form on the home page an hit submit.
10. Ensure that an error message (an alert) is displayed saying that a token is not found.
